robotoff/prediction/category/neural/keras_category_classifier_3_0/preprocessing.py

```diff
@@ -0,0 +1,272 @@
+            continue
+        # dict mapping lang to a set of expressions for a specific ingredient
+        seen: dict[str, set[str]] = defaultdict(set)
+        for field in ("names", "synonyms"):
+            for lang in ("xx", "en", "fr", "en", "es", "de", "nl", "it"):
+                names = getattr(node, field).get(lang)
+                if names is None:
+                    continue
+                if isinstance(names, str):
+                    # for 'names' field, the value is a string
+                    names = [names]
+
+                for name in names:
+                    normalized_name = fold(name.lower())
+                    if normalized_name in seen[lang] or len(normalized_name) <= 1:
+                        # Don't keep the normalized name if it already exists
+                        # or if it's length is <= 1 (to avoid false positives)
+                        # For example 'd' is a synonym for 'vitamin d'
+                        continue
+                    synonyms[(node.id, lang)].add(normalized_name)
+                    name_map[normalized_name].add((node.id, lang))
+                    seen[lang].add(normalized_name)
+
+    if add_synonym_combinations:
+        for normalized_name, keys in list(name_map.items()):
+            # get tokens from name by performing a simple whitespace
+            # tokenization
+            tokens = normalized_name.split(" ")
+            if len(tokens) <= 1:
+                continue
+
+            for current_node_id, current_lang in keys:
+                # combinations is a list of set of string, each element being
+                # a token with multiple possible synonyms.
+                # we initialize with the original tokens (if no synonym is
+                # found for any token, we will just generate the original
+                # normalized name again)
+                combinations = [{token} for token in tokens]
+                for token_idx in range(len(tokens)):
+                    token = tokens[token_idx]
+                    if token in name_map:
+                        # Lookup all ingredient IDs with the same lang that
+                        # match the normalized token string
+                        for key in (
+                            (node_id, lang)
+                            for (node_id, lang) in name_map[token]
+                            if lang == current_lang
+                        ):
+                            for synonym in synonyms[key]:
+                                combinations[token_idx].add(synonym)
+
+                for combination in itertools.product(*combinations):
+                    # generate full ingredient name using one of the combinations
+                    name = " ".join(combination)
+                    # As name_map values are sets, we're sure there are no
+                    # duplicates
+                    name_map[name].add((current_node_id, current_lang))
```

Comment:
It's a bit hard to read, but it's really a great way to get combinations :-)

We might steal this code for taxonomy editor ! Kudos.